invenio_stats/tasks.py

```diff
@@ -38,3 +43,142 @@ def process_events(event_types):
     for e in event_types:
         results.append((e, current_stats.events[e].processor.run()))
     return results
+
+
+@shared_task
+def aggregate_event(event):
+    """Calculate file downloads for today."""
+    file_download_aggregator = StatAggregator(
+        client=current_search_client,
+        event='file_download',
+        time_frame='monthly')
+    file_download_aggregator.run()
+
+
+class StatPreprocessor(object):
+    """Preprocessor class.
+
+    Subclass this class in order to provide specific statistics calculations.
+    """
+
+    def run(self):
+        """Run calculations."""
+        pass
+
+
+class StatAggregator(StatPreprocessor):
+    """Aggregator class."""
+
+    def __init__(self, client, event,
+                 time_frame, query='*'):
+        """Construct aggregator instance."""
+        self.client = client
+        self.event = event
+        self.time_frame = time_frame
+        self.query = query
+        self.aggregation_alias = '{0}-{1}'.format(self.time_frame,
+                                                  self.event)
+
+    def get_bookmark(self):
+        """Get last aggregation date."""
+        if not Index(self.aggregation_alias,
+                     using=self.client).exists():
+            oldest_index = self.get_indices_with_alias('events-stats_{}'.
+                                                       format(self.event))
+            return datetime.datetime.strptime(oldest_index,
+                                              'events-stats_{}-%Y.%m.%d'.
+                                              format(self.event)).date()
+
+        import ipdb
+        ipdb.set_trace()
+        query_bookmark = Search(using=self.client,
+                                index=self.aggregation_alias,
+                                doc_type='{0}_bookmark'.format(self.event))
+        query_bookmark.aggs.metric('latest', 'max', field='date')
+        bookmark = query_bookmark.execute()[0]
+        bookmark = datetime.datetime.strptime(bookmark.date, '%Y-%m-%d')
+        return bookmark.date()
+
+    def set_bookmark(self):
+        """Set bookmark for starting next aggregation."""
+        def _success_date():
+            bookmark = {
+                'date': datetime.date.today().isoformat()
+            }
+            yield dict(_index='{0}-{1}-{2}'.
```

Comment:
Also I don't see any error handling. One thing that should never happen is: having day 1 fail but day 2 succeed. It would mean that we would never reprocess day 1 as the latest bookmark is for day 2.